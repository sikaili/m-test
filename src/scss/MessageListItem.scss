
.MessageListItem {
    position: relative;
    margin: 1px;
    padding: 24px 32px 24px;
    background-color: #fff;

    &:hover {
        background-color: rgb(245, 245, 245);
    }

    &--active {
        background-color: rgb(202, 216, 255);
    }

    &__description {
        display: -webkit-box;
        margin-top: 8px;
        overflow: hidden;
        color: rgb(119,119,119);
        text-overflow: ellipsis;
        -webkit-line-clamp: 2; /* number of lines to show */
        -webkit-box-orient: vertical;
    }

    &__icon {
        position: absolute;
        top: 20px;
        left: 6px;
        color: #CCCCCC;

        &--active {
            color: #0074E4;
        }
    }

    &__time {
        position: absolute;
        top: 24px;
        right: 8px;
        color: #7a7a7a;

        &--active {
            color: #0074E4;
        }
    }
}