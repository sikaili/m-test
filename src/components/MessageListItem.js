import "../scss/MessageListItem.scss";

import propTypes from "prop-types";
import React from "react";

import { formatTime } from "../js/helpers";

export default function MessageListItem({
  sender, subject, description, icon, time, read, active,
}) {
  return (
    <div className={`MessageListItem ${active ? "MessageListItem--active" : ""}`}>
      <h3 className="MessageListItem__sender">{sender}</h3>
      <h4>{subject}</h4>
      <span className="MessageListItem__description">{description}</span>
      <i className={`MessageListItem__icon mypro-icon mypro-icon-${icon} ${!read ? "MessageListItem__icon--active" : ""}`} />
      <time className={`MessageListItem__time ${!read ? "MessageListItem__time--active" : ""}`} dateTime={time}>{formatTime(time)}</time>
    </div>
  );
}

MessageListItem.propTypes = {
  sender: propTypes.string,
  subject: propTypes.string,
  description: propTypes.string,
  icon: propTypes.string,
  time: propTypes.string,
  read: propTypes.bool,
  active: propTypes.bool,
};
MessageListItem.defaultProps = {
  sender: "",
  subject: "",
  description: "",
  icon: "",
  time: "",
  read: false,
  active: false,
};
