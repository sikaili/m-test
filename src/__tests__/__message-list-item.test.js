import { mount } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";

import MessageListItem from "../components/MessageListItem";

const timeString = "2021-05-10T08:16:07.049461";
const dataFormatted = "10/05/2021";
const name = "Tom";

test("renders correctly", () => {
  const component = renderer.create(
    <MessageListItem sender={name} time={timeString} />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Date is correct", () => {
  const wrapper = mount(
    <MessageListItem sender={name} time={timeString} />,
  );
  const sender = wrapper.find(".MessageListItem__sender");
  const time = wrapper.find(".MessageListItem__time");
  expect(sender && sender.text()).toBe(name);
  expect(time.text()).toBe(dataFormatted);
});

test("Display weekday when within a week", () => {
  const timeString = String(Number(new Date()) - 24 * 3600 * 1000 * 3);
  const wrapper = mount(
    <MessageListItem time={timeString} />,
  );
  const time = wrapper.find(".MessageListItem__time");
  expect(time.text()).toMatch(/[a-z]+/);
});
