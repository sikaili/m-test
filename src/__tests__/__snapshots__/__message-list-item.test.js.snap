// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="MessageListItem "
>
  <h3
    className="MessageListItem__sender"
  >
    Tom
  </h3>
  <h4>
    
  </h4>
  <span
    className="MessageListItem__description"
  >
    
  </span>
  <i
    className="MessageListItem__icon mypro-icon mypro-icon- MessageListItem__icon--active"
  />
  <time
    className="MessageListItem__time MessageListItem__time--active"
    dateTime="2021-05-10T08:16:07.049461"
  >
    10/05/2021
  </time>
</div>
`;
